### PR TITLE
[amp-list-load-more] documentation, demo, bugfix, and ui tweaks

### PR DESCRIPTION
**Pull request by cathyxz**
Part of https://github.com/ampproject/amphtml/issues/14060, https://github.com/ampproject/amphtml/issues/9423. 

1. Add more examples, in particular an endpoint that simulates pagination, # of items for load-end testing. 
2. Increase viewport preload constant from 1.5 to 3 viewports down. 
3. Minor bugfix regarding load failed state and loading state. 
4. Added `reloadCallback`, to implement button to reload last link after 404-ing.

### CODE REVIEW

File: extensions/amp-list/amp-list.md [hunk 1]
```diff
@@ -261,6 +261,35 @@ We recommend using `binding="no"` or `binding="refresh"` for faster performance.
 
 If `binding` attribute is not provided, default is `always`.
 
+## Experimental: Infinite Scroll (amp-list-load-more)
```
Comment (cathyxz) [inline extensions/amp-list/amp-list.md]: A google doc version of this has been shared internally at go/amp-list-load-more-doc for easier iteration.

File: extensions/amp-list/0.1/amp-list.js [hunk 2]
```diff
@@ -692,6 +708,29 @@ export class AmpList extends AMP.BaseElement {
...
   }
 
+  /**
+   * @private
+   */
+  setLoadMoreReload_() {
+    if (this.loadMoreOverflow_) {
+      this.mutateElement(() => {
+        this.loadMoreOverflow_.classList.toggle('amp-visible', true);
+        listen(this.loadMoreOverflow_, 'click',
```
Comment (cvializ) [inline extensions/amp-list/0.1/amp-list.js]: Should this ever be unlistened?

File: build-system/app.js [hunk 3]
```diff
@@ -1164,6 +1164,39 @@ app.get('/dist/ww(.max)?.js', (req, res) => {
   });
 });
 
+app.get('/infinite-scroll', function(req, res) {
+  const query = {req};
```
Comment (cvializ) [inline build-system/app.js]: should this be `const {query} = req`?

File: extensions/amp-list/amp-list.md (hunk 1 above)
Comment (cvializ) [reply extensions/amp-list/amp-list.md]: /cc @CrystalFaith

File: extensions/amp-list/amp-list.md [hunk 4]
```diff
@@ -261,6 +261,35 @@ We recommend using `binding="no"` or `binding="refresh"` for faster performance.
...
+
+### Attributes
+#### load-more (mandatory)
+Adding this attribute will allow amp-list (with no value) to show a “load-more” button at the end of the amp-list. The value of this attribute can be set to “auto” to trigger automatic loading more elements three viewports down for an infinite scroll effect.
+
+#### load-more-bookmark (mandatory)
+This attribute specifies an attribute in the returned data that will give the url of the next items to load. E.g. In the following sample payload, we would specify `load-more-bookmark=”next”`.
+
+```
+{ “items”: [], “next”: “https://url.to.load” }
```
Comment (cvializ) [inline extensions/amp-list/amp-list.md]: nit: You should replace these smart-quotes `“”` with regular quotes `""`, so people who copy-paste from the documentation don't need to correct them after they paste

File: build-system/app.js (hunk 3 above)
Comment (cathyxz) [reply build-system/app.js]: Oops yes. Great catch!

File: extensions/amp-list/amp-list.md (hunk 4 above)
Comment (cathyxz) [reply extensions/amp-list/amp-list.md]: Yup, awkward copy-paste from google docs.

File: extensions/amp-list/0.1/amp-list.js (hunk 2 above)
Comment (cathyxz) [reply extensions/amp-list/0.1/amp-list.js]: Yes, it should be unlistened after the load succeeds, but not if it fails. Updated the code to reflect this.

File: extensions/amp-list/0.1/amp-list.js [hunk 5]
```diff
@@ -373,13 +371,32 @@ export class AmpList extends AMP.BaseElement {
           items = items.slice(0, maxLen);
         }
         return this.scheduleRender_(/** @type {!Array} */(items), !!opt_append);
-      }, error => {
-        throw user().createError('Error fetching amp-list', error);
       });
     }
-    return fetch.catch(error => this.showFallback_(error));
+
```
[the remark this answers is not in the document]
Comment (aghassemi) [reply extensions/amp-list/0.1/amp-list.js]: > defaulting to showing load-more-button if load-more-failed is not defined

LGTM

File: extensions/amp-list/0.1/amp-list.js [hunk 6]
```diff
@@ -672,41 +691,77 @@ export class AmpList extends AMP.BaseElement {
...
+   * Called when a fetch fails under load-more. Shows the load-more-button
+   * element and triggers a reloading of the failed src on click.
+   * @private
+   */
+  setLoadMoreReload_() {
+    if (this.loadMoreButton_) {
+      this.mutateElement(() => {
+        this.loadMoreButton_.classList.toggle('amp-visible', true);
+        this.unlistenLoadMoreButton_ = listen(this.loadMoreButton_, 'click',
+            () => this.loadMoreReloadCallback_());
```
Comment (aghassemi) [inline extensions/amp-list/0.1/amp-list.js]: can `loadMoreReloadCallback` and `loadMoreCallback_` become the same thing? they are almost doing the same functionality. (at minimum the common part should be refactored)

Comment (cathyxz) [reply extensions/amp-list/0.1/amp-list.js]: Done

File: build-system/app.js [hunk 7]
```diff
@@ -1164,6 +1164,42 @@ app.get('/dist/ww(.max)?.js', (req, res) => {
...
+  const nextUrl = '/infinite-scroll?items=' +
+    numberOfItems + '&left=' + JSON.stringify(pagesLeft - 1);
+
+  if (latency) {
+    setTimeout(() => res.json({
+      'items': results,
+      'next': pagesLeft == 0 ? undefined : nextUrl,
+    }), latency);
+  } else {
+    res.json({'items': results, 'next': nextUrl});
```
[the remark this answers is not in the document]
Comment (cathyxz) [reply build-system/app.js]: Done

**Comment by cathyxz** [comment]
Merging master for the bundle size check issue.

File: build-system/app.js [hunk 8]
```diff
@@ -1164,6 +1164,42 @@ app.get('/dist/ww(.max)?.js', (req, res) => {
...
+    results.push(r);
+  }
+
+  const nextUrl = '/infinite-scroll?items=' +
+    numberOfItems + '&left=' + JSON.stringify(pagesLeft - 1);
+
+  if (latency) {
+    setTimeout(() => res.json({
+      'items': results,
+      'next': pagesLeft == 0 ? undefined : nextUrl,
```
[the remark this answers is not in the document]
Comment (cathyxz) [reply build-system/app.js]: Fixed this inline.

File: extensions/amp-list/amp-list.md (hunk 1 above)
Comment (cathyxz) [reply extensions/amp-list/amp-list.md]: Filed follow up https://github.com/ampproject/amphtml/issues/19502 assigned to @CrystalFaith 
 and myself.

File: extensions/amp-list/amp-list.md [hunk 9]
```diff
@@ -281,6 +281,35 @@ We recommend using `binding="no"` or `binding="refresh"` for faster performance.
 
 If `binding` attribute is not provided, default is `always`.
 
+## Experimental: Infinite Scroll (amp-list-load-more)
+We've introduced an experiment called `amp-list-load-more` as an implementation for pagination and infinite scroll in amp-list. This is an experimental feature, and final APIs may change.
```
Comment (CrystalFaith) [inline extensions/amp-list/amp-list.md]: nit: remove comma between "experimental feature, and final"

File: extensions/amp-list/amp-list.md [hunk 10]
```diff
@@ -281,6 +281,35 @@ We recommend using `binding="no"` or `binding="refresh"` for faster performance.
...
+
+### Attributes
+#### load-more (mandatory)
+Adding this attribute will allow amp-list (with no value) to show a “load-more" button at the end of the amp-list. The value of this attribute can be set to “auto" to trigger automatic loading more elements three viewports down for an infinite scroll effect.
```
Comment (CrystalFaith) [inline extensions/amp-list/amp-list.md]: Mind updating all mentions of `amp-list` are surrounded in ``` and tag marks?

File: extensions/amp-list/amp-list.md [hunk 11]
```diff
@@ -281,6 +281,35 @@ We recommend using `binding="no"` or `binding="refresh"` for faster performance.
...
+
+```
+{ "items": [], "next": "https://url.to.load" }
+```
+
+### Additional children of `<amp-list>`
+`<amp-list>` with the `load-more` attribute expects the following additional child elements:
+
+#### load-more-button (mandatory)
+An element containing the `load-more-button` attribute. Clicking on this button will trigger a fetch to load more elements from the url contained in the field of the data returned corresponding to the `load-more-bookmark` attribute.
```
Comment (CrystalFaith) [inline extensions/amp-list/amp-list.md]: We call it an element and a button. Should the element be `<button>` or can it be any element used as a button? If there are only specific types of elements that can be used we should list them.

**Pull request merged**